programs/bpf_loader/src/lib.rs

```diff
@@ -723,48 +861,62 @@ fn process_loader_upgradeable_instruction(
+                transaction_context,
+                upgradeable_ins_acc_idx::Upgrade::Buffer as usize,
+            )?;
 
-            // Update the ProgramData account, record the upgraded data, and zero
-            // the rest
+            // Update the ProgramData account, record the upgraded data, and zero the rest
             programdata.set_state(&UpgradeableLoaderState::ProgramData {
                 slot: clock.slot,
-                upgrade_authority_address: Some(*authority.unsigned_key()),
+                upgrade_authority_address: current_upgrade_authority_address,
             })?;
-            programdata.try_account_ref_mut()?.data_as_mut_slice()
+            programdata.get_data_mut()
                 [programdata_data_offset..programdata_data_offset + buffer_data_len]
-                .copy_from_slice(&buffer.try_account_ref()?.data()[buffer_data_offset..]);
-            programdata.try_account_ref_mut()?.data_as_mut_slice()
-                [programdata_data_offset + buffer_data_len..]
-                .fill(0);
+                .copy_from_slice(&buffer.get_data()[buffer_data_offset..]);
+            programdata.get_data_mut()[programdata_data_offset + buffer_data_len..].fill(0);
 
             // Fund ProgramData to rent-exemption, spill the rest
-
-            spill.try_account_ref_mut()?.checked_add_lamports(
-                (programdata.lamports()? + buffer.lamports()?)
-                    .saturating_sub(programdata_balance_required),
-            )?;
-            buffer.try_account_ref_mut()?.set_lamports(0);
-            programdata
-                .try_account_ref_mut()?
-                .set_lamports(programdata_balance_required);
+            {
+                let mut spill = instruction_context.try_borrow_instruction_account(
+                    transaction_context,
+                    upgradeable_ins_acc_idx::Upgrade::Spill as usize,
+                )?;
+                spill.checked_add_lamports(
+                    (programdata.get_lamports() + buffer.get_lamports())
+                        .saturating_sub(programdata_balance_required),
+                )?;
+                buffer.set_lamports(0);
+                programdata.set_lamports(programdata_balance_required);
+            }
 
             ic_logger_msg!(log_collector, "Upgraded program {:?}", new_program_id);
         }
         UpgradeableLoaderInstruction::SetAuthority => {
-            let account = keyed_account_at_index(keyed_accounts, first_instruction_account)?;
-            let present_authority =
-                keyed_account_at_index(keyed_accounts, first_instruction_account + 1)?;
-            let new_authority =
-                keyed_account_at_index(keyed_accounts, first_instruction_account + 2)
-                    .ok()
-                    .map(|account| account.unsigned_key());
-
-            match account.state()? {
+            instruction_context.check_number_of_instruction_accounts(2)?;
```

Comment:
3?

Reply:
Just 2, because `NewAuthority` is optional.